### DISCUSSION TITLE
Memory issue

### DISCUSSION THREAD

**Discussion by KiKoS0** [encode/httpx#1726] [Potential Issue]
I have a FastAPI application that has an endpoint which gets the `og:image` meta tag of html pages, it uses a `httpx.AsyncClient` to stream the response and stop reading when it finds the tag. My issue is that after a dozen requests, the memory usage grows exponentially and the application crashes. I traced it down to find out that the problem is that I'm breaking out of reading the response when i find the tag:
```python
async def getImg(url):
   og_line = None
   async with httpx.AsyncClient() as client:
        async with client.stream('GET', url) as response:
            async for chunk in response.aiter_lines():
                if "og:image" in chunk:
                    og_line = chunk
                    break
```
The memory leak stops when i comment the `break`, so i wanted to ask if this is expected to happen and that i should avoid doing that, or is this unexpected behavior?
I recreated a [minimal reproducible example](https://github.com/KiKoS0/httpx-mem-leak) so you can test it out.
I also found out that it only happens on specific websites that I'm currently dealing with. (so i included both normal and leaking links in the example)
Thanks.

**Comment by lovelydinosaur** [comment]
Any folks here want to start out by checking if you can reproduce the issue as described here?
That'd be our first step in determining if we need to do something more here.

@KiKoS0 This looks like a pretty good start. If you're able to minimise the example you've linked so that it's even more slimline (keep taking things away until you can't anymore), and include it in a comment (rather than a linked repo) then we're more likely to be able to prioritise looking into this.

**Reply by KiKoS0** [reply]
Absolutely, here you go:
```python
import httpx
import asyncio

async def getImgUrl(url):
    async with httpx.AsyncClient() as client:
        async with client.stream('GET', url) as response:
            async for chunk in response.aiter_lines():
                if "og:image" in chunk:
                    break
async def main():
    link = "https://www.autoscout24.com/offres/bmw-140-serie-1-f21-lci2-06-2017-05-2019-xdrive-340-ch-essence-bleu-0e0ef7ae-9aac-491e-e053-0100007f9bbc"

    for i in range(0,10000):
        await getImgUrl(link)

asyncio.run(main())
```

**Reply by lovelydinosaur** [reply]
Great, that's really helpful, will look into it.

Incidentally, it *really, really* makes sense to just use a single client instance, like so...

```python
import httpx
import asyncio

async def getImgUrl(client, url):
    async with client.stream('GET', url) as response:
        async for chunk in response.aiter_lines():
            if "og:image" in chunk:
                break

async def main():
    link = "https://www.autoscout24.com/offres/bmw-140-serie-1-f21-lci2-06-2017-05-2019-xdrive-340-ch-essence-bleu-0e0ef7ae-9aac-491e-e053-0100007f9bbc"

    async with httpx.AsyncClient() as client:
        for i in range(0,10000):
            await getImgUrl(client, link)

asyncio.run(main())
```

There's evidentially a lot of folks missing out on that, so we need to do a better job of highlighting the recommended usage.

It's even possible (or not) that it'll resolve the issue you're seeing. (But either way around it sounds like it needs looking into.)

**Reply by KiKoS0** [reply]
It makes perfect sense to just use one client instance (I had it set up that way originally), but unfortunately I don't think it resolves the issue.

**Reply by lovelydinosaur** [reply]
Does anyone have any good experience with how to dig out which objects are being held onto here?

/cc @encode/maintainers

**Reply by lovelydinosaur** [reply]
Might need an `aclosing()` to resolve this.

https://trio.readthedocs.io/en/stable/reference-core.html?highlight=aclosing#finalization

https://async-generator.readthedocs.io/en/latest/reference.html#contextmanagers

**Reply by johtso** [reply]
One might expect exiting the stream context manager to close the stream. I'm guessing there's a good reason this doesn't happen?

**Reply by florimondmanca** [reply]
FWIW, generally Python does what we expect when we `break` from a generator iteration and the generator's got a `finally` block:

```python
import asyncio


async def stream():
    try:
        for k in range(10):
            yield k
    finally:
        print("Finally!")


async def main():
    async for k in stream():
        print(k)
        if k == 2:
            print("Break")
            break


asyncio.run(main())
```

```console
$ python debug/client.py
0
1
2
Break
Finally!
```

**Reply by florimondmanca** [reply]
@KiKoS0 Trying your snippet and Tom's on my machine; Tom's with a single `client` stays at a stable 18.5Mo. Yours would start at around 22Mo and ramp up slowly (26Mo after ~30s).

So seems like switching to a single client solves it?

Does seems like we have some GC or "handle onto something dead" issue with lots of single-instance clients, though. Although that wouldn't be the recommended use case, I'm not sure we want that to happen anyway.

**Reply by KiKoS0** [reply]
I reran it and came to the same conclusion as you, but then I realized that the hardcoded link had expired (returns a 410 Gone).
So, strangely enough as soon as I changed it to a live one, the memory issue reappeared, even with a single client. 

@florimondmanca Could you try it again with a live link like this one?:
```python
link = "https://www.autoscout24.fr/offres/bmw-320-serie-3-touring-e91-touring-163ch-pack-m-sport-diesel-bleu-671904de-6139-4061-a451-f63bdb61de2b"
```
